### PR TITLE
Fix missing columns after query fields update

### PR DESCRIPTION
**Pull request by ranquild**
Fixes https://github.com/metabase/metabase/issues/40435
Demo https://www.loom.com/share/3e2f28fb2dd745f3b0cc20f2b62edcc7
Context https://metaboat.slack.com/archives/C01LQQ2UW03/p1710866450928429

Makes columns that are not explicitly hidden by the table viz settings visible in the table view by default. 

Usually the list of query columns is in sync with `table.columns` viz setting, but we've recently found out a case where it isn't (there is an e2e repro for this case). In this case the list of query columns now "wins" and these columns become visible until they're hidden by the user.

To reproduce:
- New -> Question -> Orders -> Pick ID and User ID -> Save
- Visualize
- Hide & show columns in the table viz settings
- Go to notebook -> Fields -> Enable Product ID -> Save without visualizing
- Visualize
- Only 2 columns would be visible by default - ID and User ID

Expected:
- Product ID should be visible unless explicitly hidden

### CODE REVIEW

File: frontend/src/metabase/visualizations/lib/settings/column.js [hunk 1]
```diff
@@ -530,34 +531,42 @@ export const buildTableColumnSettings = ({
...
-          ...setting,
-          key: getColumnSettingKey(setting),
-        }));
+        const settingIndexes = findColumnSettingIndexesForColumns(
+          cols,
+          columnSettings,
+        );
+
+        return [
+          ...columnSettings.map(setting => ({
```
Comment (ranquild) [inline frontend/src/metabase/visualizations/lib/settings/column.js:552]: existing settings

File: frontend/src/metabase/visualizations/lib/settings/column.js [hunk 2]
```diff
@@ -530,34 +531,42 @@ export const buildTableColumnSettings = ({
...
+            ...setting,
+            key: getColumnSettingKey(setting),
+          })),
+          ...cols
+            .filter((_, columnIndex) => settingIndexes[columnIndex] < 0)
```
Comment (ranquild) [inline frontend/src/metabase/visualizations/lib/settings/column.js:557]: query columns for which there is no setting

File: frontend/src/metabase/visualizations/components/settings/ChartSettingTableColumns/TableColumnPanel/TableColumnPanel.unit.spec.tsx [hunk 3]
```diff
@@ -105,16 +105,6 @@ describe("DatasetColumnSelector", () => {
     expect(items[3]).toHaveTextContent("Subtotal");
   });
 
-  it("should display columns without matching setting", () => {
```
Comment (ranquild) [inline frontend/src/metabase/visualizations/components/settings/ChartSettingTableColumns/TableColumnPanel/TableColumnPanel.unit.spec.tsx:108]: This is now not needed - the list of columns in `table.columns` is always "full"

File: e2e/test/scenarios/question/reproductions/40435-hidden-columns.cy.spec.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  restore,
+  saveQuestion,
+  visualize,
+} from "e2e/support/helpers";
+
+describe("issue 40435", () => {
+  beforeEach(() => {
+    restore();
+    cy.signInAsNormalUser();
+    cy.intercept("POST", "/api/card").as("createCard");
```
Comment (kamilmielnik) [inline e2e/test/scenarios/question/reproductions/40435-hidden-columns.cy.spec.js:17]: Unused, `saveQuestion` will wait for that endpoint call

```suggestion
```

**Review by kamilmielnik — APPROVED** [review]
Works :100:

**Pull request merged**